query/src/main/java/io/keen/client/java/KeenQueryClient.java

```diff
@@ -0,0 +1,704 @@
+                }
+            } else {
+                throw new IllegalStateException();
+            }
+        }
+
+        return new IntervalResult(intervalResult);
+    }
+
+    private static GroupByResult constructGroupByResult(ArrayList<Object> groups) {
+        Map<Group, QueryResult> groupByResult = new HashMap<Group, QueryResult>();
+
+        for (Object child : groups) {
+            if (child instanceof HashMap) {
+                HashMap<String, Object> inputMap = (HashMap<String, Object>) child;
+
+                // If this is a GroupByResult, it should have key "result", along with properties to group by.
+                if (inputMap.containsKey(KeenQueryConstants.RESULT)) {
+                    QueryResult result = null;
+                    HashMap<String, Object> properties = new HashMap<String, Object>();
+                    for (String key : inputMap.keySet()) {
+                        if (key.equals(KeenQueryConstants.RESULT)) {
+                            // there should not be intervals nested inside GroupByResult's; only
+                            // the other way around.
+                            result = constructQueryResult(inputMap.get(key), false, false);
+                        } else {
+                            properties.put(key, inputMap.get(key));
+                        }
+                    }
+
+                    Group groupBy = new Group(properties);
+                    groupByResult.put(groupBy, result);
+                } else {
+                    throw new IllegalStateException();
+                }
+            } else {
+                throw new IllegalStateException();
+            }
+        }
+
+        return new GroupByResult(groupByResult);
+    }
+
+    /**
+     * Posts a query to the server.
+     *
+     * @param params         The {@link Query} with all the required and optional arguments.
+     * @return The response from the server in the "result" map.
+     * @throws IOException If there was an error communicating with the server or
+     * an error message received from the server.
+     */
+    private Object executeHelper(Query params, Timeframe timeframe) throws IOException {
+        QueryType queryType = params.getQueryType();
+
+        if (false == params.AreParamsValid(queryType)) {
+            throw new IllegalArgumentException("Keen Query parameters are insufficient. Please check Query API docs for required arguments.");
+        }
+
+        String urlString = formatBaseURL(QueryType.getQueryType(queryType));
+
+        // Query parameter args.
+        Map<String, Object> allQueryArgs = params.ConstructQueryArgs();
+        if (timeframe != null) {
+            allQueryArgs.putAll(timeframe.constructTimeframeArgs());
+        }
+
+        Object returnVal = "";
+        URL url = new URL(urlString);
+        returnVal = publishObject(project, url, allQueryArgs);
+
+        return returnVal;
+    }
+
```

Comment:
Nit: Extra newline. I saw at least one other instance of this; I won't comment on others, and they're not a big deal, but maybe do a quick scan for them. Consistent spacing tends to make reading sources ever so slightly easier on future developers' brains.